### ISSUE TITLE
[Performance] Qwen3.5-4B (q4f16) is ~3x slower decode + ~20x slower TTFT vs Qwen3-4B on WebGPU (Transformers.js 4.0.0-next.7)

### ISSUE THREAD

**Issue by youqibing** [huggingface/transformers.js#1599]
### System Info

- **`@huggingface/transformers` version**: `4.0.0-next.7`
- **Browser**: Chrome 145.0.7632.160 (Official Build) (arm64) 
- **OS**: macOS Version 14.3.1 (Build 23D60)
- **GPU**: Apple, ANGLE Metal Renderer: Apple M2 Pro, Version 14.3.1 (Build 23D60))
- **WebGPU**: Enabled 
- **Backend**: WebGPU (`device: 'webgpu'`, `dtype: 'q4f16'`)
- **Other libraries**: None (vanilla HTML + ES module, no framework)

### Environment/Platform

- [x] Website/web-app
- [ ] Browser extension
- [ ] Server-side (e.g., Node.js, Deno, Bun)
- [ ] Desktop app (e.g., Electron)
- [ ] Other (e.g., VSCode extension)

### Environment

| Item | Value |
|------|-------|
| `@huggingface/transformers`  | `v4.0.0-next.7`  |
| Backend | WebGPU |
| Browser | Chrome 145.0.7632.160 |
| OS | macOS Version 14.3.1 (Build 23D60) |
| GPU | Apple, ANGLE Metal Renderer: Apple M2 Pro, Version 14.3.1 (Build 23D60)) |

---

### Model File Sizes (from HuggingFace Hub)

| Model | ONNX file (q4f16) | Size |
|-------|-------------------|------|
| Qwen3-4B | `model_q4f16.onnx` | ~2.4 GB |
| Qwen3.5-4B | `model_q4f16.onnx` | ~2.5 GB |

File sizes are comparable, ruling out download/IO as a factor.

---

### Description

When running an identical Japanese→English subtitle translation task
using the same quantization (`q4f16`) and the same batch pipeline,
**Qwen3.5-4B is consistently ~2–3x slower** than Qwen3-4B in net
token throughput on WebGPU.

This is unexpected given that both models have the same parameter count (4B).

---

### Suspected Root Cause

After investigation, the likely bottlenecks in Qwen3.5-4B are:

1. **Vocabulary size**: Qwen3.5-4B has **248,320 tokens** vs Qwen3-4B's
   **151,936 tokens** (~63% larger). This means the `lm_head` matmul
   on every single decode step is significantly heavier.

2. **Hybrid attention pattern**: Qwen3.5 uses a mixed layout of
   `8 × (3 sliding-window attention + 1 full attention)` layers.
   This pattern may not be fully optimized in the current ONNX export
   or ONNX Runtime WebGPU kernel path.

3. **Possible missing kernel optimization**: `MatMulNBits` /
   `GroupQueryAttention` contrib ops may not be fused/optimized
   for Qwen3.5's specific graph shape under WebGPU.

---

### Reproduction

A self-contained benchmark demo ([Qwen3 vs Qwen3.5 Subtitle Translation Benchmark.html](https://github.com/user-attachments/files/26142591/Qwen3.vs.Qwen3.5.Subtitle.Translation.Benchmark.html))
is attached. It runs both models on the **same 124-line Japanese subtitle
file**, measures net t/s, TTFT, and total time, and displays a
side-by-side comparison.

**Steps:**
1. Open the attached `benchmark.html` in Chrome 113+ with WebGPU enabled
2. Click **"▶ Run Qwen3-4B-q4f16"** — wait for completion
3. Click **"▶ Run Qwen3.5-4B-q4f16"** — wait for completion
4. Observe the speed comparison panel

> 📎 **Attached:** [Qwen3 vs Qwen3.5 Subtitle Translation Benchmark.html](https://github.com/user-attachments/files/26142591/Qwen3.vs.Qwen3.5.Subtitle.Translation.Benchmark.html) 

---

### Observed Results

| Model | Quantization | Avg net speed | Total time | Avg TTFT |
|-------|-------------|--------------|------------|----------|
| Qwen3-4B | q4f16 | **22.3 t/s** | **62.0s** | **828 ms** |
| Qwen3.5-4B | q4f16 | **14.6 t/s** | **289.6s** | **16152 ms** |

**Qwen3.5-4B is approximately `3x` slower** than Qwen3-4B.

---

### ⚠️ Critical Finding: TTFT Anomaly

The TTFT difference is far more dramatic than the throughput difference:

| Model       | Avg TTFT  | Ratio  |
|-------------|-----------|--------|
| Qwen3-4B    | 828 ms    | 1×     |
| Qwen3.5-4B  | **16,152 ms** | **~19.5×** |

This suggests the bottleneck is primarily in the **prefill/prompt-encoding
stage**, not just the decode loop. A ~19x TTFT regression with only a
~1.5x throughput regression points to a potential issue in:

- The ONNX graph structure for Qwen3.5's hybrid attention during prefill
- WebGPU shader compilation / pipeline caching for Qwen3.5-specific ops
- Possible O(n²) complexity in the sliding-window attention implementation
  during the initial prompt pass

---

### Expected Behavior

Given the same parameter count and quantization level, the two models
should have comparable inference throughput. A ~63% larger vocabulary
alone does not fully explain a 2–3x slowdown — there may be additional
ONNX graph or WebGPU dispatch inefficiencies specific to Qwen3.5.

---

### Additional Context

- Both models loaded successfully with no errors
- The slowdown is consistent across multiple batches (not a cold-start issue)
- Thinking mode was **disabled** (`<think>\n\n</think>` prepended) to ensure
  fair comparison of raw decode speed
- Tested with batch sizes of 5, 10, and 15 lines — slowdown ratio is consistent
- **The TTFT anomaly (~19.5× slower) is the most significant signal** and likely
  indicates a prefill-stage regression rather than a pure decode throughput issue
- Wall-clock total time is ~4.7× longer (62s vs 289.6s), consistent with
  per-batch TTFT overhead dominating total runtime
- Hardware: Apple M2 Pro (Metal backend via ANGLE) — WebGPU via Chrome 145
- No console errors or warnings during either run

**Comment by xenova** [comment]
Hi 👋 thanks for the testing! Most likely related to https://github.com/microsoft/onnxruntime/pull/27780

**Comment by youqibing** [comment]
> Hi 👋 thanks for the testing! Most likely related to [microsoft/onnxruntime#27780](https://github.com/microsoft/onnxruntime/pull/27780)

I'm not entirely sure whether these two issues are related. After upgrading from Transformer.js 3.8.1 to 4.0.0-next.7, I didn't notice any significant improvement in Whisper's performance on WebGPU. However, I did observe a noticeable drop in translation speed for Qwen3.5 compared to Qwen3. Could this suggest that the support for Qwen3.5 in Transformer.js 4.0.0 still has some room for optimization?

**Comment by xenova** [comment]
> I'm not entirely sure whether these two issues are related. After upgrading from Transformer.js 3.8.1 to 4.0.0-next.7, I didn't notice any significant improvement in Whisper's performance on WebGPU.

The PR isn't specific to whisper, it is more focused on specific parameters/configuration of attention nodes

**Comment by youqibing** [comment]
> > I'm not entirely sure whether these two issues are related. After upgrading from Transformer.js 3.8.1 to 4.0.0-next.7, I didn't notice any significant improvement in Whisper's performance on WebGPU.
> 
> The PR isn't specific to whisper, it is more focused on specific parameters/configuration of attention nodes

Thanks for linking [#27780](https://github.com/microsoft/onnxruntime/pull/27780)!

That makes sense — the TTFT regression (~19.5×) I observed is likely tied to how Qwen3.5's hybrid attention layers (3:1 sliding-window + full attention) are being dispatched through the current WebGPU kernel path, rather than a pure decode throughput issue.

A few follow-up questions:

1.Is the FlashAttention optimization in #27780 expected to cover Qwen3.5's specific attention configuration (e.g., sliding-window attention during prefill)?
2.Is there anything I can do to help — such as running benchmarks on my M2 Pro once the fix lands, or testing intermediate builds?
3.Is the bottleneck more on the onnxruntime side or the transformers.js ONNX export side for Qwen3.5?
Happy to provide more profiling data if useful. Thanks again for looking into this!

**Comment by kokroo** [comment]
+1
Experiencing the same issue.

**Comment by abgnydn** [comment]
This is fixed in a different repo. `onnx-community/Qwen3.5-4B-ONNX-OPT`, the re-export @xenova
pushed on 2026-04-22 and linked as `new_version` on the original card, does not have the problem.
It needs transformers.js 4.2.0 or later, for the two reasons at the bottom of this comment. The 0.8B
and 2B exports have the same `-OPT` pair.

I am posting the mechanism because neither this issue nor the `-OPT` card says what the swap
changes or why.

Prefill on the decoder graph, run directly through onnxruntime-web on synthetic inputs. M2 Max,
Chrome, onnxruntime-web 1.26.0-dev, 252 positions, two runs, each a median of 3.

| | prefill |
|---|---:|
| `onnx-community/Qwen3.5-4B-ONNX` | 13.00 / 12.99 s |
| `onnx-community/Qwen3.5-4B-ONNX-OPT` | 0.85 / 0.93 s |

That is a bare decoder `session.run()`, not an end-to-end transformers.js TTFT, so it does not
compare directly against the 16 s you measured. I have not published the harness for it. The
graph counts below are the part you can check in a minute.

The old export is slow because its 24 gated-DeltaNet layers are each an ONNX `Scan`, a sequential
loop that runs one prompt position per iteration.

```python
# pip install onnx huggingface_hub
# hf download onnx-community/Qwen3.5-4B-ONNX onnx/decoder_model_merged_q4f16.onnx --local-dir .
# Same path in -OPT. 1.4 MB old, 0.9 MB OPT. Weights live in separate .onnx_data files and are
# not needed here, so fetch the single file rather than cloning (the LFS blobs are ~40 GB).
import onnx, collections
m = onnx.load('onnx/decoder_model_merged_q4f16.onnx', load_external_data=False)
ops = collections.Counter()
def walk(g):
    for n in g.node:
        ops[n.op_type] += 1
        for a in n.attribute:
            if a.g and a.g.node: walk(a.g)
walk(m.graph)
print(ops['Scan'], ops['GroupQueryAttention'], ops['LinearAttention'])
# old: 24 8 0      OPT: 0 8 24
```

All 24 sit in the `else_branch` of a per-layer `If` whose condition is `.../gdn/is_decode`, so
only prefill runs them. onnxruntime-web has no WebGPU kernel for `Scan`, so the loop control runs
on CPU while its body dispatches to the GPU once per position, in each of the 24 layers.

I measured the cost by replacing every `Scan` in the old decoder with shape-preserving `Identity`
nodes and re-running in the same session. This was a separate session from the table above, so
the baselines differ slightly. Prefill went 13.99 s to 1.24 s in one run and 13.71 s to 0.95 s in
another, so the recurrence is about 91% and 93% of it.

The OPT export replaces each `Scan` with a fused `LinearAttention` node, and carries 24
`CausalConvWithState` nodes alongside them. The WebGPU EP has kernels for both, so the layer
stays on the GPU. OPT also adds `num_logits_to_keep`, so prefill returns `[1, 1, 248320]` instead
of `[1, 252, 248320]` and skips building a 125 MB logits tensor.

@youqibing asked whether microsoft/onnxruntime#27780 covers this. It covers part of it. That PR
optimizes the FlashAttention prefill path. It was Apple-gated as merged on 2026-05-14, extended
to NVIDIA the same day by microsoft/onnxruntime#28511, and generalised to any device without
subgroups by microsoft/onnxruntime#28520 four days later. All three first shipped in ORT 1.27,
which is later than the 1.26.0-dev build transformers.js pins today. 8 of the 32 layers here are
gated attention and would use it. The other 24 are gated-DeltaNet, and that is where the prefill
time went.
microsoft/onnxruntime#27996 (merged 2026-04-09) added the WebGPU kernels for `LinearAttention`
and `CausalConvWithState`, and the `-OPT` re-export emits them.

On the version floor, which has two gates rather than one. 4.0.0-next.7, the version in this
issue, pins onnxruntime-web `1.25.0-dev.20260307`. That build predates both
microsoft/onnxruntime#27907 (2026-04-07, CPU and CUDA kernels) and microsoft/onnxruntime#27996
(2026-04-09, the WebGPU ones), so neither op is registered in it and session creation fails
rather than falling back to CPU: `com.microsoft:LinearAttention(-1) is not a registered
function/op`.

4.1.0 pins `1.26.0-dev.20260410`, which has both ops, and still cannot run `-OPT`. It sizes the
empty conv cache from the config rather than the graph: `getCacheShapes` in `src/configs.js`
returns `past_conv.i = [batch_size, 8192, linear_conv_kernel_dim]`, and `-OPT`'s own
`config.json` still carries `linear_conv_kernel_dim: 4` from the repo it was duplicated from,
while the `-OPT` graph declares `past_conv.i` as `[batch_size, 8192, 3]`. The run fails with
`Got invalid dimensions for input: past_conv.0 ... index: 2 Got: 4 Expected: 3`.

4.2.0 replaced `getCacheShapes` with `getCacheNames` plus `resolveCacheShape` over
`session.inputMetadata`, so the shape comes from the graph and the 3 is honoured. That makes
4.2.0 the floor. It shipped 2026-04-22, about seven hours before the `-OPT` upload. The `next`
tag is still 4.0.0-next.11, which pins a 2026-03-27 build and is also too early.

Scope of what I checked: prefill only. I have not compared logits or tokens between the two
exports, and I have not measured decode, which is the other half of what you reported.

The `new_version` banner is on the old card, but this issue does not mention it and the `-OPT`
card is frontmatter only, so anyone arriving here still lands on the slow repo. I can open a PR
on the `-OPT` card with the mechanism and the version floor if that is useful.
